core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -611,15 +620,31 @@ private[spark] class TaskSetManager(
 
     // If no executors have registered yet, don't abort the stage, just wait.  We probably
     // got here because a task set was added before the executors registered.
-    if (executors.nonEmpty) {
+    if (executorsByHost.nonEmpty) {
       // take any task that needs to be scheduled, and see if we can find some executor it *could*
       // run on
-      pendingTask.foreach { taskId =>
-        if (executors.forall(executorIsBlacklisted(_, taskId))) {
-          val execs = executors.toIndexedSeq.sorted.mkString("(", ",", ")")
-          val partition = tasks(taskId).partitionId
-          abort(s"Aborting ${taskSet} because task $taskId (partition $partition)" +
-            s" has already failed on executors $execs, and no other executors are available.")
+      pendingTask.foreach { indexInTaskSet =>
+        // try to find some executor this task can run on.  Its possible that some *other*
+        // task isn't schedulable anywhere, but we will discover that in some later call,
+        // when that unschedulable task is the last task remaining.
+        val blacklistedEverywhere = executorsByHost.forall { case (host, execs) =>
+          val nodeBlacklisted = blacklist.isNodeBlacklisted(host) ||
```

Comment:
Add a comment above here with something like "Check if the task can run on the node"